include/openthread/srp_client.h

```diff
@@ -0,0 +1,489 @@
+ *    OT_ERROR_INVALID_ARGS     : The provided service structure is invalid (e.g., bad service name or `otSrpTxtEntry`).
+ *    OT_ERROR_NO_BUFS          : Insufficient buffer to prepare or send the update message.
+ *
+ * Note that in case of any failure, the client continues the operation, i.e. it prepares and (re)transmits the SRP
+ * update message to the server, after some wait interval. The retry wait interval starts from the minimum value and
+ * is increased by the growth factor every failure up to the max value (please see configuration parameter
+ * `OPENTHREAD_CONFIG_SRP_CLIENT_MIN_RETRY_WAIT_INTERVAL` and the related ones for more details).
+ *
+ *
+ * @param[in] aContext          A pointer to an arbitrary context (provided when callback was registered).
+ * @param[in] aError            The error (see above).
+ * @param[in] aHostInfo         A pointer to host info.
+ * @param[in] aService          The head of linked-list containing all services (excluding the ones removed). NULL if
+ *                              the list is empty.
+ * @param[in] aRemovedServices  The head of linked-list containing all removed services. NULL if the list is empty.
+ *
+ */
+typedef void (*otSrpClientCallback)(void *                     aContext,
+                                    otError                    aError,
+                                    const otSrpClientHostInfo *aHostInfo,
+                                    const otSrpClientService * aServices,
+                                    const otSrpClientService * aRemovedServices);
+
+/**
+ * This function starts the SRP client operation.
+ *
+ * SRP client will prepare and send "SRP Update" message to the SRP server once all the following conditions are met:
+ *
+ *  - The SRP client is started - `otSrpClientStart()` is called.
+ *  - Host name is set - `otSrpClientSetHostName()` is called.
+ *  - At least one host IPv6 address is set - `otSrpClientSetHostName()` is called.
+ *  - At least one service is added - `otSrpClientAddService()` is called.
+ *
+ * It does not matter in which order these functions are called. When all conditions are met, the SRP client will
+ * wait for a short delay before preparing an "SRP Update" message and sending it to server. This delay allows for user
+ * to add multiple services and/or IPv6 addresses before the first SRP Update message is sent (ensuring a single SRP
+ * Update is sent containing all the info).
+ *
+ * @param[in] aInstance        A pointer to the OpenThread instance.
+ * @param[in] aServerSockAddr  The socket address (IPv6 address and port number) of the SRP server.
+ * @param[in] aCallback        The callback which is used to notify events and changes. Can be NULL if not needed.
+ * @param[in] aContext         An arbitrary context used with @p aCallback.
+ *
+ * @retval OT_ERROR_NONE       SRP client operation started successfully.
+ * @retval OT_ERROR_ALREADY    SRP client is already started and running.
+ *
```

Comment:
We can have a UDP socket open/connect failure `OT_ERROR_FAILED`. Will add it.

But I don't think invalid args can be currently returned (in public API we intentionally don't check for non-null-ness of the pointers, since it adds a lot of code and increases code-size with little practical impact)).